pkg/pillar/cmd/zedagent/handlemetrics.go

```diff
@@ -528,6 +531,50 @@ func publishMetrics(ctx *zedagentContext, iteration int) {
 	}
 	log.Tracef("DirPaths in persist, elapse sec %v", time.Since(startPubTime).Seconds())
 
+	// ZFS metrics
+	if vault.ReadPersistType() == types.PersistZFS {
+		zpoolList, err := libzfs.PoolOpenAll()
+		if err != nil {
+			log.Errorf("get zpool list failed %v", err)
+		} else {
+			for _, zpool := range zpoolList {
+				rStorageMetric := new(zmet.StorageMetric)
+				rStorageMetric.PoolName = zpool.Properties[libzfs.PoolPropName].Value
+				rStorageMetric.StorageState = zfs.GetZfsDeviceStatusFromStr(zpool.Properties[libzfs.PoolPropHealth].Value)
+				rStorageMetric.CountVolume, err = zfs.GetZfsCountVolume(zpool.Properties[libzfs.PoolPropName].Value)
+				if err != nil {
+					log.Errorf("get count volume failed %v", err)
```

Comment:
In this case will the reported metrics be useful? Do we need to flag in the API that we had a failure?

Reply:
I agree here because in this case, we read the module file in sysfs and if this file is not there, then the problems will be much worse than the lack of statistics. But this should not affect the code below in any way, since it should correctly handle problems further.

Reply:
I think that the lack of data or a mismatch with the real state of affairs will make it clear anyway.
But we can add some bool in the next iteration of the API, the value of which will indicate any problems in collecting data, not only for zfs for example.

Reply:
ok